### ISSUE TITLE
Switch to python-jose/jwcrypto?

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#8322]
We originally wrote https://github.com/certbot/josepy because in ~2014 we couldn't find an implementation of JOSE in Python that worked for our needs.

Since then, https://github.com/mpdavis/python-jose has popped up and has become one of the most downloaded Python packages according to https://pythonwheels.com/.

Should we transition Certbot's code to use this library and deprecate `josepy`? I haven't looked into the library enough to decide for sure we should do that, however, I think it'd be a positive thing if we're able to retire some of our own projects because they're no longer needed. We could instead contribute to and use the more popular project and focus on other work.

To flag a couple things I noticed about `python-jose`:

* It supports multiple backends including `cryptography` so we wouldn't need a new crypto dependency.
* According to [their README as of writing this](https://github.com/mpdavis/python-jose/tree/82bd8aa1987fb574830c73f2631fbf7be22fa3d9#cryptographic-backends), it currently pulls in unnecessary dependencies, but it looks like they want to change that eventually.
* It doesn't yet seem to be packaged in distros like Ubuntu.

**Comment by ohemorange** [comment]
My main concern is about maintenance. How fast are they to accept PRs? How quickly do they respond to security incidents? Who are these people and will they suddenly stop supporting it in the near future?

> It doesn't yet seem to be packaged in distros like Ubuntu.

How would we handle this? Get it packaged? Vendorize it? Have our packages pip install it (would that even work?)?

**Comment by bmw** [comment]
> My main concern is about maintenance. How fast are they to accept PRs? How quickly do they respond to security incidents? Who are these people and will they suddenly stop supporting it in the near future?

I don't know the answers to any of that yet. I think the fact the core crypto dependency they recommend and use is `cryptography` which we already use and trust is a good sign as is the fact it was in the top 350 most commonly installed packages from PyPI in the past year with nearly 25 million installs according to https://hugovk.github.io/top-pypi-packages/, but more research is definitely needed before switching and I agree the things you mentioned are important to look into.

> How would we handle this? Get it packaged? Vendorize it? Have our packages pip install it (would that even work?)?

Partially due to us taking more control over our packaging, I was imagining we wouldn't do much here. We'd include the change in our changelog and it'd be nice to reach out to our Debian and Fedora/EPEL maintainers to personally flag this for them, but I otherwise think we'd just start requiring the dependency and distros would have to package it if they want to continue packaging Certbot. I'd be surprised if it was a major issue for any of them though.

I'll also flag that I think us doing this would require a new major version of certbot/acme since I think `josepy` objects are part of our API.

**Comment by ohemorange** [comment]
This all sounds good. The easiest way to find those answers is probably just to message them and ask, I couldn't find lots of team info on their github page, though they do seem to be regularly making changes and putting out releases.

**Comment by bmw** [comment]
I agree. Another thing I'm personally a little interested in just as another signal of health is who is currently relying on this library? Somebody is driving those download numbers.

**Comment by alexzorin** [comment]
The documentation is extremely bare.

**Comment by commonism** [comment]
josepy is not great wrt. documentation or test-coverage, python-jose may do better in testing but documentation is not an improvement. I guess something like https://github.com/spulec/moto (… easily mock out tests based on AWS infrastructure …) using python-jose for Amazon cognitoidp may drive the download numbers.

https://github.com/latchset/jwcrypto does better in test-coverage and documentation.

In certbot (python-)acme messages rely on the TypedJSONObjectWithFields provided by josepy, so replacing josepy will affect python-acme and I do not see acme message de/serialisation getting any better with either python-jose or jwcrypto when using different decoding mechanisms in acme / jose.

**Comment by bmw** [comment]
Ooo I just got around to looking at jwcrypto.

More research would need to be done, but I think it may be promising. It's based on cryptography by default (with no other supported backend) and has maintainers from Red Hat and the Python core team.

**Comment by adferrand** [comment]
Hello !

I think we need also to talk about additional elements here to embrace the full picture of what implies a move from `josepy` to another library. Indeed `josepy` is doing fulfill to different purposes:
* it is a JOSE implementation is discussed before, and a competitor of `jwcrypto` and `python-jose` in that regard
* but it is also a JSON data processing library, giving solutions for JSON serialization/deserialization with data binding, modeling and schema validation

So in my opinion, moving from `josepy` requires to find a solution for this second part also.

If we want to stay with a class-based approach, here are some possibilities:
* switch to the native `dataclasses` solution with custom validation and serialization/deserialization
* switch to the native `dataclasses` and a third party based on `dataclass` for JSON processing (eg. https://pypi.org/project/dataclasses-json/)
* switch to a fully-fledged solution like `pydantic` (https://docs.pydantic.dev/latest/)

Alternatively, we give up the class-based approach to go to a set of dedicated methods responsible for serialization/deserialization and data validation.

**Comment by adferrand** [comment]
The problem is that the these two purposes are strongly coupled in `josepy`, making the shift of the JOSE approach significantly harder, because you have to deal in parallel with the JSON processing approach, that is structural to certbot code.

We will have to deal with a temporary additional complexity in a way or another. A first step is certainly to vendor the current `josepy` code into Certbot project to make it private, free of non backward compatibles changes and hassle of handling releases of the library.

**Comment by adferrand** [comment]
Looking at both libraries, from JSON data processing perspective, `jwcrypto` looks more promising. Indeed it offers low-level serialization/deserialization routines: these routines can be the base to construct a mapping to the `TypedJSONObjectWithFields` structure, then to one of the other solutions I mentioned above for instance. I can start seeing a bottom-up approach and a migration path from there.

So all things equal otherwise, I am more in favor of `jwcrypto`.

**Comment by akx** [comment]
(Bringing my comment https://github.com/certbot/josepy/issues/70#issuecomment-2310061535 here.)

[PyJWT](https://pyjwt.readthedocs.io/en/stable/) is also an option for JW* primitives.

Based on PyPIStats:

* [PyJWT: 194 million downloads/month](https://pypistats.org/packages/pyjwt)
* [python-jose: 11 million downloads/month](https://pypistats.org/packages/python-jose)
* [jwcrypto: 3.7 million downloads/month](https://pypistats.org/packages/jwcrypto)

**Comment by ohemorange** [comment]
@akx, looks like PyJWT includes only JWT and not JOSE, no?

**Comment by akx** [comment]
> @akx, looks like PyJWT includes only JWT and not JOSE, no?

What do you think is missing? Based on eg. https://github.com/jpadilla/pyjwt/issues/942 it can be a replacement for at least python-jose...

**Comment by ohemorange** [comment]
Working on a design doc with considerations and options here: https://docs.google.com/document/d/1UQUeEF5NsytWNZtsiQ3qBUsgrZvSGVKIu0N-Z88Ze5w/edit?usp=sharing

**Comment by commonism** [comment]
Hi,

pypi download numbers are not a hard reference, these numbers are not driven by "production usage" but unit testing.
For pyjwt - google-auth-library-python is such dependency.

python-jose is (already) considered unmaintained and evolving to a security problem for all projects depending on it.

pydantic mixing nullable with required/optional was addressed in 2.0.
msgspec may worth looking into if you want to avoid the pydantic-core dependency.

**Comment by ohemorange** [comment]
That's a good point @commonism! As you'll see in the document, `python-jose` is no longer under consideration, and `joserfc` is currently the top choice over `pyjwt` despite having lower monthly downloads.

That's really helpful to hear about `pydantic` 2.0, and will certainly take a look at `msgspec` as well.

**Comment by ohemorange** [comment]
Based on discussion, we're going to go with option 1) continue to use `josepy`, deprecating usages of `PyOpenSSL` in favor of `cryptography`, so I'm closing this issue.
